src/Bedrock.Framework.Experimental/Protocols/Kafka/KafkaClientConnectionContext.cs

```diff
@@ -0,0 +1,131 @@
+{
+    /// <summary>
+    /// Encapsulates a client Kafka connection that channels requests to an underlying connection
+    /// in the order that they were written; and reads responses in the same order.
+    /// </summary>
+    public class KafkaClientConnectionContext : IKafkaClientConnectionContext
+    {
+        private readonly Channel<(object, IKafkaMessageDescriptor, TaskCompletionSource<object>)> _requestChannel =
+            Channel.CreateUnbounded<(object, IKafkaMessageDescriptor, TaskCompletionSource<object>)>(
+                new UnboundedChannelOptions()
+                {
+                    SingleReader = true,
+                    SingleWriter = false
+                });
+
+        private readonly ConcurrentQueue<(int, IKafkaMessageDescriptor, TaskCompletionSource<object>)> _requestQueue =
+            new ConcurrentQueue<(int, IKafkaMessageDescriptor, TaskCompletionSource<object>)>();
+
+        private readonly IMessageWriter<KafkaRequest> _requestWriter;
+        private readonly IMessageReader<KafkaResponse> _responseReader;
+
+        private int _correlationId = 0;
+
+        /// <summary>
+        /// Initializes a new <see cref="KafkaClientConnectionContext"/>.
+        /// </summary>
+        /// <param name="requestWriter">An <see cref="IMessageWriter{KafkaRequest}"/> that can write Kafka requests to
+        /// a buffer.</param>
+        /// <param name="responseReader">An <see cref="IMessageReader{KafkaResponse}"/> that can read Kafka responses from
+        /// a buffer.</param>
+        public KafkaClientConnectionContext(IMessageWriter<KafkaRequest> requestWriter, IMessageReader<KafkaResponse> responseReader)
+        {
+            _requestWriter = requestWriter;
+            _responseReader = responseReader;
+        }
+
+        /// <inheritdoc/>
+        public Task StartAsync(ConnectionContext connectionContext, CancellationToken cancellationToken = default)
+        {
+            _ = WriteRequestsAsync(connectionContext, cancellationToken);
+            _ = ReadResponsesAsync(connectionContext, cancellationToken);
+
+            return Task.CompletedTask;
+        }
+
+        /// <inheritdoc/>
+        public async Task<TResponseSchema> SendAsync<TRequestSchema, TResponseSchema>(
+            TRequestSchema request,
+            IKafkaMessageDescriptor messageDescriptor,
+            CancellationToken cancellationToken = default)
+        {
+            var tcs = new TaskCompletionSource<object>();
+
+            _requestChannel.Writer.TryWrite((request, messageDescriptor, tcs));
+
+            return (TResponseSchema)await tcs.Task;
+        }
+
+        private async Task WriteRequestsAsync(ConnectionContext connectionContext, CancellationToken cancellationToken = default)
+        {
+            var writer = connectionContext.CreateWriter();
+
+            while (await _requestChannel.Reader.WaitToReadAsync(cancellationToken).ConfigureAwait(false))
+            {
```

Comment:
Could this be replaced with
`await foreach(var request in _requestChannel.Reader.ReadAllAsync())`

Reply:
My understanding is that the `await foreach` will be async for each request while the current implementation (`.WaitToReadAsync` and `.TryRead`) will only go async while it is truly waiting for network bytes/requests to show up but be synchronous in the case where there are multiple requests available to be read without having to take an async hit for each one.

Reply:
Guess what ReadAllAsync does ;)

https://github.com/dotnet/runtime/blob/master/src/libraries/System.Threading.Channels/src/System/Threading/Channels/ChannelReader.netcoreapp.cs#L19

Reply:
The IAsync enumerable uses a single value task source and calls it over and over so there should be zero (or max 1 allocation at the start of the loop ) overhead or difference using the ReadAllAsync and the code is much more readable